### PR TITLE
Consistent output shape from `get_image_features`

### PR DESCRIPTION
**Pull request by zucchini-nlp**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46405)

# What does this PR do?

as per title, branches off from https://github.com/huggingface/transformers/pull/45783

The pooled image (OR video, NOT audio yet) feature output will now always satisfy three cond, where image features can be a list or a 3D tensor:
1) len(image_features) = len(input_images) # NOTE: or num_videos, not number of total video frames
2) image_features[0].ndim == 2
3) image_features[0].shape == {{actual seq length of this image, LM hidden size}}

Could make it complete BC and return the "correct" output when a certain flag is passed, but for now decided to update directly smaller utility fn in a breaking way

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
run-slow: aya_vision, cohere2_vision, deepseek_ocr2, gemma4, paddleocr_vl, qwen2_5_omni, qwen3_omni_moe

File: src/transformers/models/aya_vision/modular_aya_vision.py [hunk 1]
```diff
@@ -135,7 +135,10 @@ def get_image_features(
                 hs_pool = [hs[:, 1:] for hs in hs_pool]
             selected_image_feature = torch.cat(hs_pool, dim=-1)
 
-        image_outputs.pooler_output = self.multi_modal_projector(selected_image_feature)
+        pooler_output = self.multi_modal_projector(selected_image_feature)
+        image_outputs.pooler_output = pooler_output.reshape(
+            selected_image_feature.shape[0], -1, self.config.text_config.hidden_size
+        )
```
Comment (vasqu) [inline src/transformers/models/aya_vision/modular_aya_vision.py:142]: Does this reshape not affect the downstream forward?

File: src/transformers/models/deepseek_ocr2/modular_deepseek_ocr2.py [hunk 2]
```diff
@@ -1045,7 +1044,7 @@ def forward(
             image_features = self.get_image_features(
                 pixel_values, pixel_values_local, num_local_patches, return_dict=True
             ).pooler_output
-            image_features = image_features.to(inputs_embeds.device, inputs_embeds.dtype)
+            image_features = torch.cat(image_features, dim=0).to(inputs_embeds.device, inputs_embeds.dtype)
```
Comment (vasqu) [inline src/transformers/models/deepseek_ocr2/modular_deepseek_ocr2.py:1047]: any reason the cat is not directly in get img features?

File: src/transformers/models/gemma4/modular_gemma4.py [hunk 3]
```diff
@@ -1987,7 +1993,7 @@ def forward(
         # Merge text and images
         if pixel_values is not None:
             image_features = self.get_image_features(pixel_values, image_position_ids, return_dict=True).pooler_output
-            image_features = image_features.to(inputs_embeds.device, inputs_embeds.dtype)
+            image_features = torch.cat(image_features, dim=0).to(inputs_embeds.device, inputs_embeds.dtype)
```
Comment (vasqu) [inline src/transformers/models/gemma4/modular_gemma4.py:2064]: probably happens a few times so leaving it for a last time

File: src/transformers/models/gemma4/modular_gemma4.py [hunk 4]
```diff
@@ -1853,7 +1853,13 @@ def get_image_features(
...
         )
         last_hidden_state = vision_outputs.last_hidden_state
-        vision_outputs.pooler_output = self.embed_vision(inputs_embeds=last_hidden_state)
+        pooler_output = self.embed_vision(inputs_embeds=last_hidden_state)
+
+        output_length = pixel_values.shape[-2] // (self.config.vision_config.pooling_kernel_size**2)
+        k_squared = int((image_position_ids.shape[1] // output_length) ** 0.5) ** 2
+        non_pad_mask = (image_position_ids != -1).all(dim=-1)
+        split_sizes = (non_pad_mask.sum(dim=-1) // k_squared).tolist()
+        vision_outputs.pooler_output = torch.split(pooler_output, split_sizes)
```
Comment (vasqu) [inline src/transformers/models/gemma4/modular_gemma4.py:1924]: Ig this is similar to how the processor calculates the number of images/videos? Would be nice if we could have a cross ref

File: src/transformers/models/paddleocr_vl/modeling_paddleocr_vl.py [hunk 5]
```diff
@@ -1182,7 +1182,8 @@ def get_image_features(
         vision_outputs = self.visual(pixel_values=pixel_values, grid_thw=image_grid_thw, **kwargs)
         image_embeds = vision_outputs.last_hidden_state
         image_embeds = self.projector(image_embeds, image_grid_thw)
-        vision_outputs.pooler_output = image_embeds
+        split_sizes = (image_grid_thw.prod(-1) // self.config.vision_config.spatial_merge_size**2).tolist()
+        vision_outputs.pooler_output = torch.split(image_embeds, split_sizes)
```
Comment (vasqu) [inline src/transformers/models/paddleocr_vl/modeling_paddleocr_vl.py:1176]: Hmm ok this will happen a few times; adjusting my previous comment a bit. Would it make sense to have a global function for this that is used across processor and model as here?

**Review by vasqu — COMMENTED** [review]
Some initial comments. I think it's overall fine to break but I would want @tomaarsen's opinion as well since he worked quite a bit on this and uses it downstream so just wanna make sure we coordinate

Other than that some design questions
- Do we not cat in get image features, why?
- We sometimes have to calculate the expected number of tokens produced by the vision tower; would it make sense to align across processor/modeling by having some global function?

**Comment by vasqu** [comment]
We also need to add :rotating_light: for sure

File: src/transformers/models/aya_vision/modular_aya_vision.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/aya_vision/modular_aya_vision.py:142]: nope, if used in core VLM it won't affect. I reshaped inside VLMs forward when needed

File: src/transformers/models/deepseek_ocr2/modular_deepseek_ocr2.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/deepseek_ocr2/modular_deepseek_ocr2.py:1047]: then we have a vague 2D tensor of (total_seq_length, dim) while I am moving towardsa unified output. So the output will show clearly how many images there are and make it easy to get an embeddings of `image_i`

File: src/transformers/models/gemma4/modular_gemma4.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/gemma4/modular_gemma4.py:1924]: nah, it is taken from Gemma4Pooling layer but yeah, will add a ref

File: src/transformers/models/paddleocr_vl/modeling_paddleocr_vl.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/models/paddleocr_vl/modeling_paddleocr_vl.py:1176]: haha, this happens for all qwen-based models which is indeed a lot atm. Wdym by a global fn since I'm not sure it's worth for a one-liner `torch.split`

**Comment by zucchini-nlp** [comment]
> Do we not cat in get image features, why?
Answered in comments, to get clearly split encodings per image so we don't have to guess. It'll make to easy to get embeddings of image-i and expand/crop/do whatever you need 

> We sometimes have to calculate the expected number of tokens produced by the vision tower; would it make sense to align across processor/modeling by having some global function?

You mean gemma4 (which I also replied is based on modeling, not processor)? Otherwise we always pass the number of grids (as in qwen) or just split based on input pixel shape

**Comment by zucchini-nlp** [comment]
Oh yeah, Tom requested this in the past for `last_hidden_states` which I personally didn't need for generation 😅  Leaving it open, and working only on `pooler_output` for now

File: src/transformers/models/paddleocr_vl/modeling_paddleocr_vl.py (hunk 5 above)
Comment (vasqu) [reply src/transformers/models/paddleocr_vl/modeling_paddleocr_vl.py:1176]: I thought maybe a static function that can be both consumed by the model / processor to get the expected number here. But yea indeed it's not too worthwhile for now

File: src/transformers/models/deepseek_ocr2/modular_deepseek_ocr2.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/deepseek_ocr2/modular_deepseek_ocr2.py:1047]: Ah true, yea that makes sense

**Comment by vasqu** [comment]
Yea I think we are mostly fine then, still want to wait on @tomaarsen if he has any opinions on this and maybe @remi-or using it for CB(?)

**Comment by zucchini-nlp** [comment]
yeah, let's wait for ppl affected. I also commented on CB to let Remi know :)

**Comment by remi-or** [comment]
In general, anything that makes shapes more static or standardize inputs is good for CB, so this is great!  

The only problem that this could create is if `torch.split` is called with a number of splits that cannot be determined beforehand, which would not be CUDA graphable. Eventually, we will want CUDA graphs around the encoder.
But, since multimodal for CB has not landed yet, and the first iteration of that won't have CUDA graphs around the encoder, maybe we solve that problem when we get to it. Then, maybe an option to return the inputs before the split would do the trick.

**Comment by tomaarsen** [comment]
I updated my ST test suite a bit [here](https://github.com/huggingface/sentence-transformers/pull/3829), and then ran it in conjunction with this branch: no issues. The code also looks solid from my perspective, i.e. more standardized is more better 😄 

@remi-or 's concern is fair though.

- Tom Aarsen

**Comment by zucchini-nlp** [comment]
> if torch.split is called with a number of splits that cannot be determined beforehand

indeed, the number if splits would be dynamic each call depending on input image. Even though it is part of `inputs`, the split will cause graphbreak. And now that I think about it, we're interfering with Ilyas' efforts to make VLMs exportable 😿 

@vasqu I'll tag you after checking in with exporters and finding a better way to make everyone happy

**Comment by vasqu** [comment]
I think we can revisit this after #41992? I'm not sure but I don't think it would interfere with Ilyas work as we forward the inputs once and capture the necessary metadata based on that - so yea it should be fine on that side

**Comment by zucchini-nlp** [comment]
@vasqu , let's merge? I rebased main and fixed new models

File: tests/models/gemma4/test_modeling_gemma4.py [hunk 6]
```diff
@@ -602,6 +614,109 @@ def count_calls(*args, **kwargs):
             _ = model(inputs_embeds=inputs_embeds)
             self.assertEqual(counter["call_count"], 1)
 
+    @parameterized.expand([True, False, None])
+    def test_get_image_features_output(self, return_dict: bool | None):
```
Comment (vasqu) [inline tests/models/gemma4/test_modeling_gemma4.py:618]: small docstring to why we overrode pls

**Review by vasqu — APPROVED** [review]
One small comment, but
1. Let's add 🚨 as this is technically still breaking (in a good way)
2. Get everything on board cc @tomaarsen @IlyasMoutawwakil if you could have a final check as well so we don't break exporters and ST 
3. Fix fast CI 😬 

But yes, I think it's mergable. Just want everyone to be aware that it happens

File: src/transformers/models/kimi_k25/modeling_kimi_k25.py
Comment (tomaarsen) [inline src/transformers/models/kimi_k25/modeling_kimi_k25.py]: `get_video_features` just calls `get_image_features`, so we need to make the same modifications that we did for images for videos here. 
Another potential issue is that `get_image_features` seems to pool away the temporal dimension of video inputs. Is that expected?

File: src/transformers/models/gemma4/modular_gemma4.py
Comment (tomaarsen) [inline src/transformers/models/gemma4/modular_gemma4.py]: Video still just returns the unsplit tensor, is that fine?

**Review by tomaarsen — COMMENTED** [review]
I did some tests for Sentence Transformers, and found that this PR shouldn't affect it. ST only really uses the `get_..._features` methods when the `forward` does not contain a `last_hidden_state` (e.g. CLIP, SigLIP, etc.), as then the `get_..._features` methods must be used to get per-modality outputs. 

For all of these modern architectures, the `forward` accepts any combination of modalities already, so I don't have to rely on the `get_..._features` methods.

I did write two small comments though, based on my and Fable's review of the work.

**Comment by zucchini-nlp** [comment]
Thanks for review, fixing that and hopefully merging later in the day
